crates/wasmtime/src/instance.rs

```diff
@@ -145,57 +168,55 @@ impl Instance {
-        let mut i = Instantiator::new(store, module, imports)?;
+        let mut i = Instantiator::new(&mut store.opaque(), module, imports)?;
         loop {
-            if let Some((instance, items)) = i.step()? {
-                if instance.handle.module().start_func.is_some() {
+            let step = i.step(&mut store.opaque())?;
+            if let Some((id, instance)) = step {
+                let start = store.instance(id).module().start_func;
+                if let Some(start) = start {
                     store
-                        .on_fiber(|| Instantiator::start_raw(&instance))
+                        .on_fiber(|store| Instantiator::start_raw(store, id, start))
                         .await??;
                 }
-                if let Some(items) = items {
-                    break Ok(Instance::from_wasmtime(&items, store));
+                if let Some(instance) = instance {
+                    break Ok(instance);
                 }
             }
         }
     }
 
-    pub(crate) fn from_wasmtime(handle: &RuntimeInstance, store: &Store) -> Instance {
-        Instance {
-            items: handle.clone(),
-            store: store.clone(),
-        }
+    pub(crate) fn from_wasmtime(handle: RuntimeInstance, store: &mut StoreOpaque) -> Instance {
+        Instance(store.store_data_mut().insert(handle))
     }
 
     /// Returns the type signature of this instance.
-    pub fn ty(&self) -> InstanceType {
+    pub fn ty(&self, store: impl AsContext) -> InstanceType {
+        let store = store.as_context();
+        let items = &store[self.0];
         let mut ty = InstanceType::new();
-        for export in self.exports() {
-            ty.add_named_export(export.name(), export.ty());
+        for (name, item) in items.iter() {
+            ty.add_named_export(name, item.ty(&store));
         }
         ty
     }
 
-    pub(crate) fn wasmtime_export(&self) -> &RuntimeInstance {
-        &self.items
+    pub(crate) fn items<'a>(&self, store: &'a StoreData) -> &'a RuntimeInstance {
+        &store[self.0]
     }
 
-    /// Returns the associated [`Store`] that this `Instance` is compiled into.
-    ///
-    /// This is the [`Store`] that generally serves as a sort of global cache
-    /// for various instance-related things.
-    pub fn store(&self) -> &Store {
-        &self.store
+    pub(crate) fn comes_from_same_store(&self, store: &StoreOpaque) -> bool {
+        store.store_data().contains(self.0)
     }
 
     /// Returns the list of exported items from this [`Instance`].
-    pub fn exports<'instance>(
-        &'instance self,
-    ) -> impl ExactSizeIterator<Item = Export<'instance>> + 'instance {
-        self.items.iter().map(move |(name, item)| {
-            let extern_ = unsafe { Extern::from_wasmtime_export(item, &self.store) };
-            Export::new(name, extern_)
-        })
+    pub fn exports<'a, T: 'a>(
+        &'a self,
+        store: impl Into<StoreContextMut<'a, T>>,
```

Comment:
Note that this doesn't use `AsContextMut` so the liftetime can be explicitly closed over in the returned iterator

Reply:
It's also worth pointing out that `Mut` isn't technically needed here, yet. I realized that today, though, if you instantiate something we'll create a separate hash map in the wasmtime crate for all the exports (and their corresponding `Extern` value). 

That feels inefficient and not necessary since the hash map is already stored within `Arc<Module>`. I'd like to improve this in the future so instantiation simply records the `InstanceId` and nothing else, and then lookup/fetching here actually mutates the store to create a new `Func` (or whatever it is) on-demand rather than doing it all up front.

This means that in the future mutability may be required, so I went ahead and added it here.